keanu-project/src/test/java/io/improbable/keanu/algorithms/variational/KLDivergenceTest.java

```diff
@@ -0,0 +1,137 @@
+package io.improbable.keanu.algorithms.variational;
+
+import com.google.common.collect.Iterables;
+import io.improbable.keanu.algorithms.NetworkSamples;
+import io.improbable.keanu.algorithms.mcmc.MetropolisHastings;
+import io.improbable.keanu.network.BayesianNetwork;
+import io.improbable.keanu.network.NetworkState;
+import io.improbable.keanu.tensor.dbl.DoubleTensor;
+import io.improbable.keanu.vertices.VertexId;
+import io.improbable.keanu.vertices.dbl.DoubleVertex;
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.ConstantDoubleVertex;
+import io.improbable.keanu.vertices.dbl.probabilistic.GaussianVertex;
+import io.improbable.keanu.vertices.dbl.probabilistic.ProbabilisticDouble;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.lessThan;
+import static org.junit.Assert.assertEquals;
+import org.junit.Test;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class KLDivergenceTest {
+
+    @Test
+    public void klDivergenceIsZeroIfPAndQAreIdentical() {
+        double identicalProb = Math.log(0.5);
+
+        QDistribution qDist = mock(QDistribution.class);
+        when(qDist.getLogOfMasterP(any(NetworkState.class))).thenReturn(identicalProb);
+        NetworkSamples samples = createNetworkSamplesWithOneVertexAndOneSample(identicalProb);
+
+        assertEquals(0., KLDivergence.compute(qDist, samples), 1e-6);
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void throwsExceptionIfPProbIsNotZeroButQIs() {
+        QDistribution q = mock(QDistribution.class);
+        when(q.getLogOfMasterP(any(NetworkState.class))).thenReturn(Double.NEGATIVE_INFINITY);
+        NetworkSamples samples = createNetworkSamplesWithOneVertexAndOneSample(Math.log(0.5));
+
+        KLDivergence.compute(q, samples);
+    }
+
+    @Test
+    public void klDivergenceReturnsZeroIfPIsZeroButQIsNot() {
+        QDistribution q = mock(QDistribution.class);
+        when(q.getLogOfMasterP(any(NetworkState.class))).thenReturn(Math.log(0.5));
```

Comment:
you could pull this out into a method:

    private QDistribution mockThatReturnsConstantValue(double identicalProb) {
        QDistribution q = mock(QDistribution.class);
        when(q.getLogOfMasterP(any(NetworkState.class))).thenReturn(identicalProb);
        return q;
    }